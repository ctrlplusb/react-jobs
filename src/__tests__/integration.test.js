/* eslint-disable react/prop-types */

import React from 'react'
import { renderToStaticMarkup } from 'react-dom/server'
import asyncBootstrapper from 'react-async-bootstrapper'
import { mount } from 'enzyme'
import { resolveAfter, rejectAfter } from '../../tools/tests/helpers'
import { withJob, createJobContext, JobProvider } from '../'

const workTime = 10

function ResultRenderer({ jobResult, children }) {
  return (
    <div>
      {jobResult}
      {children && <div>{children}</div>}
    </div>
  )
}
const ErrorComponent = ({ error }) => <div>{error ? error.message : null}</div>
const error = new Error('Oh noes!')
error.stack = 'fake stack'

const createComponents = () => ({
  Hello: withJob({ work: () => resolveAfter(workTime, 'Hello') })(
    ResultRenderer,
  ),
  World: withJob({ work: () => resolveAfter(workTime, 'World') })(
    ResultRenderer,
  ),
  Goodbye: withJob({ work: () => resolveAfter(workTime, 'Goodbye') })(
    ResultRenderer,
  ),
  Fail: withJob({
    work: () => rejectAfter(workTime, error),
    ErrorComponent,
  })(ResultRenderer),
})

const createApp = (context, state) => {
  const { Hello, World, Goodbye, Fail } = createComponents()
  return (
    <JobProvider jobContext={context} rehydrateState={state}>
      <div>
        <Hello>
          <World />
        </Hello>
        <Goodbye />
        <Fail />
      </div>
    </JobProvider>
  )
}

describe('integration tests', () => {
  it('render server and client', () => {
    // we have to delete the window to emulate a server only environment
    const windowTemp = global.window
    delete global.window

    // Server side render
    const jobContext = createJobContext()
    const serverApp = createApp(jobContext)
    return (
      asyncBootstrapper(serverApp)
        .then(() => {
          const serverString = renderToStaticMarkup(serverApp)
          // SNAPSHOT 1
          expect(serverString).toMatchSnapshot()
          const stateForClient = jobContext.getState()
          // SNAPSHOT 2
          expect(stateForClient).toMatchSnapshot()
          // Restore the window and attach the state to the "window" for the
          // client
          global.window = windowTemp
          return { serverHTML: serverString, stateForClient }
        })
        .then(({ serverHTML, stateForClient }) => {
          // "Client" side render...

          const clonedState = {
            jobs: Object.assign({}, stateForClient.jobs),
          }

          const clientApp = createApp(undefined, stateForClient)
          const clientRenderWrapper = mount(clientApp)
          // SNAPSHOT 3
          expect(clientRenderWrapper).toMatchSnapshot()

          // Also validate that it matches server result. Need to return
          // a new result as data is removed after rehydration.
          expect(
            renderToStaticMarkup(createApp(undefined, clonedState)),
          ).toEqual(serverHTML)

          return clientRenderWrapper
        })
        // Now give the client side components time to resolve
        .then(
          clientRenderWrapper =>
            new Promise(resolve =>
              setTimeout(() => resolve(clientRenderWrapper), 100),
            ),
        )
        // Now a full render should have occured on client
        // SNAPSHOT 4
        .then(clientRenderWrapper =>
          expect(clientRenderWrapper).toMatchSnapshot(),
        )
    )
  })
})
