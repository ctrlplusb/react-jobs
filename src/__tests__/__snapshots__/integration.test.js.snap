// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration tests render server and client 1`] = `"<div><div>Hello<div><div>World</div></div></div><div>Goodbye</div><div>Oh noes!</div></div>"`;

exports[`integration tests render server and client 2`] = `
Object {
  "jobs": Object {
    "1": Object {
      "data": "Hello",
    },
    "2": Object {
      "data": "World",
    },
    "3": Object {
      "data": "Goodbye",
    },
    "4": Object {
      "error": Object {
        "message": "Oh noes!",
        "stack": "fake stack",
      },
    },
  },
}
`;

exports[`integration tests render server and client 3`] = `
<JobProvider
  jobContext={null}
  rehydrateState={
    Object {
      "jobs": Object {},
    }
  }
>
  <div>
    <WithJob(ResultRenderer)>
      <ResultRenderer
        jobResult="Hello"
      >
        <div>
          Hello
          <div>
            <WithJob(ResultRenderer)>
              <ResultRenderer
                jobResult="World"
              >
                <div>
                  World
                </div>
              </ResultRenderer>
            </WithJob(ResultRenderer)>
          </div>
        </div>
      </ResultRenderer>
    </WithJob(ResultRenderer)>
    <WithJob(ResultRenderer)>
      <ResultRenderer
        jobResult="Goodbye"
      >
        <div>
          Goodbye
        </div>
      </ResultRenderer>
    </WithJob(ResultRenderer)>
    <WithJob(ResultRenderer)>
      <ErrorComponent
        error={
          Object {
            "message": "Oh noes!",
            "stack": "fake stack",
          }
        }
      >
        <div>
          Oh noes!
        </div>
      </ErrorComponent>
    </WithJob(ResultRenderer)>
  </div>
</JobProvider>
`;

exports[`integration tests render server and client 4`] = `
<JobProvider
  jobContext={null}
  rehydrateState={
    Object {
      "jobs": Object {},
    }
  }
>
  <div>
    <WithJob(ResultRenderer)>
      <ResultRenderer
        jobResult="Hello"
      >
        <div>
          Hello
          <div>
            <WithJob(ResultRenderer)>
              <ResultRenderer
                jobResult="World"
              >
                <div>
                  World
                </div>
              </ResultRenderer>
            </WithJob(ResultRenderer)>
          </div>
        </div>
      </ResultRenderer>
    </WithJob(ResultRenderer)>
    <WithJob(ResultRenderer)>
      <ResultRenderer
        jobResult="Goodbye"
      >
        <div>
          Goodbye
        </div>
      </ResultRenderer>
    </WithJob(ResultRenderer)>
    <WithJob(ResultRenderer)>
      <ErrorComponent
        error={
          Object {
            "message": "Oh noes!",
            "stack": "fake stack",
          }
        }
      >
        <div>
          Oh noes!
        </div>
      </ErrorComponent>
    </WithJob(ResultRenderer)>
  </div>
</JobProvider>
`;
